cloud/grouping.py

```diff
@@ -0,0 +1,130 @@
+from cloudferrylib.os.compute import nova_compute
+
+
+class Grouping(object):
+    def __init__(self, config, group_file, cloud_id):
+        self.config = config
+        self.group_config = self.get_group_conf(group_file)
+        resources = {'identity': keystone.KeystoneIdentity,
+                     'network': neutron.NeutronNetwork,
+                     'compute': nova_compute.NovaCompute}
+        self.cloud = cloud.Cloud(resources, cloud_id, config)
+
+        self.network = self.cloud.resources['network']
+        self.compute = self.cloud.resources['compute']
+        self.identity = self.cloud.resources['identity']
+
+        self.groups = {}
+
+    @staticmethod
+    def get_group_conf(group_file):
+        with open(group_file) as gf:
+            group_conf = yaml.load(gf)
+        return group_conf
+
+    @staticmethod
+    def write_groups(file_name, groups):
+        with open(file_name, 'w') as gr:
+            yaml.dump(groups, gr)
+
+    def group(self):
+        group_by = self.group_config.pop('group_by')
+
+        for step, grouping in enumerate(group_by):
+            groups = self.group_by(grouping, step)
+            if not step:
+                self.groups = groups
+
+        self.make_users_group(self.groups)
+
+        self.write_groups(self.config.migrate.group_file_path, self.groups)
+
+    def group_by(self, target, step):
+        group_rules_map = {'tenant': self.group_nested_tenant,
+                           'network': self.group_nested_network,
+                           }
+        group_func = group_rules_map.get(target)
+
+        if not group_func:
+            raise RuntimeError("There is no such grouping option. Use 'tenant'"
+                               " or 'network' values in the 'group_by' section"
+                               " of the group config file")
+
+        if not step:
+            return group_func()
+        else:
+            self.walk(self.groups, group_func)
+            return self.groups
+
+    def group_nested_tenant(self, instances_list=None):
+        groups = {}
+        tenant_list = self.identity.get_tenants_list()
+        instances_list = instances_list if instances_list else None
+
+        for tenant in tenant_list:
+            vms_list = self.compute.get_instances_list(
+                tenant_id=tenant.id,
+                user_list=instances_list)
+            groups[str(tenant.id)] = [str(vm.id) for vm in vms_list]
+
+        return groups
+
+    def group_nested_network(self, instances_list=None):
+        groups = {}
+        network_list = self.network.get_networks_list()
+        instances_list = instances_list if instances_list else None
+
+        for network in network_list:
+            vms_list = self.compute.get_instances_list(
+                network_name=network['name'],
+                user_list=instances_list)
+
+            groups[str(network['name'])] = [str(i.id) for i in vms_list]
+
+        return groups
```

Comment:
`group_nested_network` and `group_nested_tenant` are almost identical, please keep DRY and refactor to decrease the amount of repeated code in conjunction with refactoring done in `nova_compute.py`.